### ISSUE TITLE
🐛 [Bug] Donations slider broken when missing translations

### ISSUE THREAD

**Issue by Tirokk** [Ocelot-Social-Community/Ocelot-Social#2356]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-11-25T18:26:54Z" title="Monday, November 25th 2019, 7:26:54 pm +01:00">Nov 25, 2019</time>_

---

## :bug: Bugreport

If you are using a language that has untranslated phrases/words for the donations section, it doesn't show the amount or any fallback language like English

### Steps to reproduce the behavior
1. log into your account
2. switch to Italian
3. see that the Donations slider is broken
4. ...
5. Profit

### Expected behavior

It is quite important that this information is always displayed, even if it's in a fallback language

![Screenshot from 2019-11-25 19-24-08](https://user-images.githubusercontent.com/26943915/69567376-87cdcf80-0fb9-11ea-9e0e-5360ebf04e9a.png)

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2019-11-28T14:05:02Z" title="Thursday, November 28th 2019, 3:05:02 pm +01:00">Nov 28, 2019</time>_

---


Im an absolutly newcomer and this is my first contribution - if possible. I can reproduce this behaviour frontend but where can I see the code?

**Comment by Tirokk** [comment]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-13T09:10:48Z" title="Friday, December 13th 2019, 10:10:48 am +01:00">Dec 13, 2019</time>_

---


oh, hey @erhardtstiefel I didn't see this until now.. sorry.. I think you found where the translations go, no?
I guess this will be a problem for any missing translations, so I think we need to have a backup, like English
Here is a screen shot of more broken code on production 
![Screenshot from 2019-12-13 10-08-28](https://user-images.githubusercontent.com/26943915/70788274-cffa3980-1d90-11ea-9c40-df81557c4d4c.png)

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2019-12-13T09:27:52Z" title="Friday, December 13th 2019, 10:27:52 am +01:00">Dec 13, 2019</time>_

---


Hello @mattwr18 , yes this is the first issue I picked up myself to start with, but @roschaefer had other plans to bring me in a shape where I could contribute to this fascinating project. As soon as possible I'll look now at this issue and if I've problems I'll ask you. Ok?

**Comment by Tirokk** [comment]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-13T09:29:41Z" title="Friday, December 13th 2019, 10:29:41 am +01:00">Dec 13, 2019</time>_

---


great, some more feedback for you is that it looks like the only one it doesn't work for is Italian
all others have a fallback of English when the translation is missing... perhaps Italian isn't registered properly?

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2019-12-13T14:03:49Z" title="Friday, December 13th 2019, 3:03:49 pm +01:00">Dec 13, 2019</time>_

---


I added the missing italian verbs just for the display of "donations" but I see a lot more verbs missing in it.json. Does it make sense to translate and  fill in everything which is to be found in en.json ?

**Comment by Tirokk** [comment]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-13T17:07:43Z" title="Friday, December 13th 2019, 6:07:43 pm +01:00">Dec 13, 2019</time>_

---


That would be awesome, yes @erhardtstiefel

**Comment by Tirokk** [comment]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-12-13T17:07:54Z" title="Friday, December 13th 2019, 6:07:54 pm +01:00">Dec 13, 2019</time>_

---


That would be awesome, yes @erhardtstiefel

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2019-12-14T18:39:48Z" title="Saturday, December 14th 2019, 7:39:48 pm +01:00">Dec 14, 2019</time>_

---


@mattwr18 Can I use this issue as a reference for a PR that contains all my additions to it.json or do I have to create a new issue?

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2019-12-18T17:17:37Z" title="Wednesday, December 18th 2019, 6:17:37 pm +01:00">Dec 18, 2019</time>_

---


1.) I have added all missing texts in it.json. Therefore I translated the texts from en.json with DeepL into Italian. 
2.) How can I test my work?
3.) I found a difference in "moderation": { "name": "Moderation", "reports": { ....
between en.json and it.json. It looks like there was code added in en.json but forgotten to syncronise it to it.json. What can I do?

**Comment by Tirokk** [comment]
> [<img alt="stale[bot]" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/apps/stale) **Authored by [stale[bot]](https://github.com/apps/stale)**
_<time datetime="2020-02-16T18:14:06Z" title="Sunday, February 16th 2020, 7:14:06 pm +01:00">Feb 16, 2020</time>_

---


This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2020-02-17T06:10:37Z" title="Monday, February 17th 2020, 7:10:37 am +01:00">Feb 17, 2020</time>_

---


Which type of activity do you else expect? I asked some questions and didn't get any answer.

**Comment by Tirokk** [comment]
> [<img alt="stale[bot]" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/apps/stale) **Authored by [stale[bot]](https://github.com/apps/stale)**
_<time datetime="2020-04-17T06:51:08Z" title="Friday, April 17th 2020, 8:51:08 am +02:00">Apr 17, 2020</time>_

---


This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

**Comment by Tirokk** [comment]
> [<img alt="rbeer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-04-18T12:30:22Z" title="Saturday, April 18th 2020, 2:30:22 pm +02:00">Apr 18, 2020</time>_

---


Hello Erhard! :)

> Can I use this issue as a reference for a PR that contains all my additions to it.json or do I have to create a new issue?

Yes, just reference this one.

> 2.) How can I test my work?

You will have to get a local development instance running. There is a [guide](https://docs.human-connection.org/human-connection/installation) for that. This all still depends a bit on what OS you are running. Feel free to ask, when you have any problems! 

> 3.) I found a difference in "moderation": { "name": "Moderation", "reports": { ....
between en.json and it.json. It looks like there was code added in en.json but forgotten to syncronise it to it.json. What can I do?

Since you have added the other translations, you could do the same with those missing and add them.

Having translations is obviously the better choice over fallback strings, but let me address the problem.

**tl;dr** _i18n finds empty strings in [it.json](https://github.com/Human-Connection/Human-Connection/blob/master/webapp/locales/it.json) and uses them. The defined fallback language 'en' isn't used on empty strings, but only on `undefined`, i.e. when the key is missing. -> Don't use empty strings, unless you explicitly want to. Be aware that this has layout side effects!_

The translation flow is:

0. (during setup): define `fallbackLanguage` to be 'en'
1. figure out `language` to use ('de', 'en', 'it', ...)
2. look up `translatedString` for `key` in `language`
3. _is `translatedString` defined?_
3.1 _yes_ - use it
3.2 _no_ - use `translatedString` from `fallbackLanguage`

Which explains the missing keys in 'it'.
'de' is the only file that has the same amount of lines - implying amount of `key`s - as 'en'.
![locale-lines](https://user-images.githubusercontent.com/3411649/79637086-8a091400-817c-11ea-9ee7-09c529eb0f47.png)

What makes 'it' different and causes this bug is that 'it' is the only language that has empty strings as `translatedString`s.
![emptyString-it-only](https://user-images.githubusercontent.com/3411649/79637268-46170e80-817e-11ea-91db-6e09b3003350.png)

The empty strings then get used, instead of replaced by the value from `fallbackLanguage`. Leaving empty containers behind, that sometimes inflate completely:
![login_en](https://user-images.githubusercontent.com/3411649/79637437-f0436600-817f-11ea-98d5-da3e1f5a8abc.png)
![login_it](https://user-images.githubusercontent.com/3411649/79637435-efaacf80-817f-11ea-9bfc-c23306608e91.png)

This is what it looks like when you remove the keys `forgotPassword` and `register` from it.json; now with fallback english:
![login_it_fixed](https://user-images.githubusercontent.com/3411649/79637929-10c0ef80-8183-11ea-9da0-9b992182e509.png)


Takeaway: Don't use empty strings! Either add a translation or leave the key undefined to use the fallback language (i.e. 'en'glish)

**Comment by Tirokk** [comment]
> [<img alt="rbeer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-04-19T18:03:30Z" title="Sunday, April 19th 2020, 8:03:30 pm +02:00">Apr 19, 2020</time>_

---


@erhardtstiefel why did you close this? We need to get rid of those empty strings in it.json:
https://github.com/Human-Connection/Human-Connection/blob/32e5801d20ddcb3c56407dc7f02f00c8c316337a/webapp/locales/it.json#L36-L40

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2020-04-26T17:38:49Z" title="Sunday, April 26th 2020, 7:38:49 pm +02:00">Apr 26, 2020</time>_

---


Sorry, but as I mentioned I do not know how to fix it although I'm interested to know how to do it.

**Comment by Tirokk** [comment]
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2020-04-26T18:18:36Z" title="Sunday, April 26th 2020, 8:18:36 pm +02:00">Apr 26, 2020</time>_

---


I think this issue can be closed @rbeer and @erhardtstiefel because the issue is about the Donations slider that is no longer visible for users, nor is it clear if it will ever be used again.
There might be another issue that there are missing translations, but as far as I remember @roschaefer made to where there is a fallback language, which is English, I believe. Obviously, we would like to have everything translated in Italian, but that relies on Italian language users adding them in lokalise or in the code base.

**Comment by Tirokk** [comment]
> [<img alt="rbeer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/rbeer) **Authored by [rbeer](https://github.com/rbeer)**
_<time datetime="2020-04-27T08:17:16Z" title="Monday, April 27th 2020, 10:17:16 am +02:00">Apr 27, 2020</time>_

---


@mattwr18 Maybe rename it, because the issue is not minor, yet easy to fix.
The missing (or rather empty) strings often let their component's layout collapse; entire buttons, links, paragraphs just disappearing.

@erhardtstiefel 
You basically already did the major bulk of the work.
Take a look at 9ec166e! Simply removing all those keys that have empty strings as values solves the issue, because then the fallback language is used.
Since you have already translated many of the missing keys, your version of it.json is obviously preferred.

So what you would have to do is just remove all the empty strings you have left and open a pull request for that one file (it.json).

There are some keys with the value `null`
https://github.com/Human-Connection/Human-Connection/blob/771fe49da66a520e457c416310b605b0ba29c72e/webapp/locales/it.json#L601
I don't know how our i18n library handles that, but if you don't have a translation for those keys, it's safe to remove them, just like the empty strings. Worst case `null` means "use fallback language", which is the same as removing them.

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2020-04-27T10:49:42Z" title="Monday, April 27th 2020, 12:49:42 pm +02:00">Apr 27, 2020</time>_

---


Scusate, non sono italiano e non sono un programmatore. Volevo solo aiutare a modificare un numero che è stato dichiarato per i principianti.
Sorry, I'm not Italian and I'm not a programmer. I just wanted to help editing an issue that was declared for beginners.
Tut mir leid, ich bin kein Italiener und kein Programmierer. Ich wollte nur helfen ein issue zu bearbeiten, das für Anfänger deklariert war.

**Comment by Tirokk** [comment]
> [<img alt="stale[bot]" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/apps/stale) **Authored by [stale[bot]](https://github.com/apps/stale)**
_<time datetime="2020-06-26T12:34:00Z" title="Friday, June 26th 2020, 2:34:00 pm +02:00">Jun 26, 2020</time>_

---


This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs. Thank you for your contributions.

**Comment by Tirokk** [comment]
> [<img alt="erhardtstiefel" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/erhardtstiefel) **Authored by [erhardtstiefel](https://github.com/erhardtstiefel)**
_<time datetime="2020-06-26T14:50:55Z" title="Friday, June 26th 2020, 4:50:55 pm +02:00">Jun 26, 2020</time>_

---


Hi, anybody still ongoing with Human Connection?

**Comment by Tirokk** [comment]
This should be solved by #4357

**Comment by Tirokk** [comment]
This should be solved by #4357
